File: spec/integration/message_metadata_spec.rb
require 'rails_helper'

RSpec.describe 'Message Metadata', type: :request do
  let(:account) { create(:account) }
  let(:inbox) { create(:inbox, account: account) }
  let(:conversation) { create(:conversation, account: account, inbox: inbox) }
  let(:agent) { create(:user, account: account, role: :agent) }
  let(:headers) { agent.create_new_auth_token }

  before do
    create(:inbox_member, inbox: inbox, user: agent)
  end

  describe 'POST /api/v1/accounts/{account.id}/conversations/{conversation.id}/messages' do
    context 'when message contains urgency keywords' do
      it 'detects urgency and issue type' do
        post "/api/v1/accounts/#{account.id}/conversations/#{conversation.display_id}/messages",
             params: { content: 'URGENT: System is not working, need immediate help' },
             headers: headers

        expect(response).to have_http_status(:success)
        json_response = response.parsed_body
        expect(json_response['content_attributes']['urgency']).to eq('high')
        expect(json_response['content_attributes']['issue_type']).to eq('technical')
      end
    end

    context 'when message contains medium priority keywords' do
      it 'detects medium urgency' do
        post "/api/v1/accounts/#{account.id}/conversations/#{conversation.display_id}/messages",
             params: { content: 'Important: Need help with billing question' },
             headers: headers

        expect(response).to have_http_status(:success)
        json_response = response.parsed_body
        expect(json_response['content_attributes']['urgency']).to eq('medium')
        expect(json_response['content_attributes']['issue_type']).to eq('billing')
      end
    end

    context 'when message contains no urgency keywords' do
      it 'does not add urgency metadata' do
        post "/api/v1/accounts/#{account.id}/conversations/#{conversation.display_id}/messages",
             params: { content: 'Just wanted to say hello' },
             headers: headers

        expect(response).to have_http_status(:success)
        json_response = response.parsed_body
        expect(json_response['content_attributes']['urgency']).to be_nil
        expect(json_response['content_attributes']['issue_type']).to be_nil
      end
    end
  end
end

File: app/javascript/dashboard/components/widgets/conversation/specs/MessageMetadata.spec.js
import { shallowMount } from '@vue/test-utils';
import MessageMetadata from '../MessageMetadata.vue';

describe('MessageMetadata.vue', () => {
  it('renders correctly with urgency and issue type', () => {
    const wrapper = shallowMount(MessageMetadata, {
      props: {
        contentAttributes: {
          urgency: 'high',
          issue_type: 'technical',
        },
      },
      global: {
        mocks: {
          $t: msg => msg,
        },
      },
    });

    expect(wrapper.find('.message-metadata').exists()).toBe(true);
    expect(wrapper.find('.urgency').exists()).toBe(true);
    expect(wrapper.find('.badge').text()).toBe('HIGH');
    expect(wrapper.find('.value').text()).toBe('technical');
  });

  it('does not render when no metadata is present', () => {
    const wrapper = shallowMount(MessageMetadata, {
      props: {
        contentAttributes: {},
      },
    });

    expect(wrapper.find('.message-metadata').exists()).toBe(false);
  });

  it('renders only urgency when issue type is missing', () => {
    const wrapper = shallowMount(MessageMetadata, {
      props: {
        contentAttributes: {
          urgency: 'medium',
        },
      },
      global: {
        mocks: {
          $t: msg => msg,
        },
      },
    });

    expect(wrapper.find('.urgency').exists()).toBe(true);
    expect(wrapper.find('.badge').text()).toBe('MEDIUM');
    expect(wrapper.find('.value').exists()).toBe(false);
  });

  it('renders only issue type when urgency is missing', () => {
    const wrapper = shallowMount(MessageMetadata, {
      props: {
        contentAttributes: {
          issue_type: 'billing',
        },
      },
      global: {
        mocks: {
          $t: msg => msg,
        },
      },
    });

    expect(wrapper.find('.urgency').exists()).toBe(false);
    expect(wrapper.find('.value').text()).toBe('billing');
  });
});
